### PR TITLE
Feat/patpass commerce

### PR DESCRIPTION
**Pull request by TheYakuza**
Add transbank patpass comercio SDK
complete inscription method with its exceptions and response
they are all at __patpass_comercio__ folder

### CODE REVIEW

**Comment by a1b4** [comment]
You should create PR by functionality form example one for create, one for commit, etc. review 62 files is so hard and require too much time.
You description does not details what do this PR what means add Patpass comercio Sdk?

**Comment by TheYakuza** [comment]
@eugenioLopezRamos Done!, I've changed the branch as you recommend

File: lib/patpass/patpass_comercio/Inscription.php [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+use Transbank\Patpass\PatpassComercio\Exceptions\InscriptionFinishException;
+use Transbank\Patpass\PatpassComercio\Exceptions\InscriptionStatusException;
+
+
+class Inscription
+{
+    const INSCRIPTION_START_ENDPOINT = 'restpatpass/v1/services/patInscription';
+    const INSCRIPTION_FINISH_ENDPOINT = 'restpatpass/v1/services/patInscription/$TOKEN$';
+
+    public static function start(
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/Inscription.php]: this is extremely awkward, is there no way to  make this not have over 9000 parameters while still maintaining a similar style to the rest of the sdk? eg: summing up a lot of the user's personal info into a `associative array` with that data?

File: lib/patpass/patpass_comercio/exceptions/InscriptionFinishException.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+namespace Transbank\Patpass\PatpassComercio\Exceptions;
+
+use Transbank\Patpass\Exceptions\PatpassException;
+
+class InscriptionFinishException extends PatpassException
+{
+
+    public function __construct($message = self::DEFAULT_MESSAGE, $code = 0, $previous = null)
+    {
+        parent::__construc($message, $code, $previous);
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/exceptions/InscriptionFinishException.php]: `construc` is a bug that we should fix in all of the exceptions that have it

File: lib/patpass/patpass_comercio/inscription/InscriptionFinishResponse.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ * @package Transbank\Patpass\PatpassComercio
+ *
+ */
+
+
+namespace Transbank\Patpass\PatpassComercio;
+
+
+class InscriptionFinishResponse
+{
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/inscription/InscriptionFinishResponse.php]: You use this passing it a response json, but there is no constructor/setters etc

File: lib/patpass/patpass_comercio/inscription/InscriptionStatusResponse.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ */
+
+
+namespace Transbank\Patpass\PatpassComercio;
+
+
+class InscriptionStatusResponse
+{
+    public $status;
+    public$urlVoucher;
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/inscription/InscriptionStatusResponse.php]: there's a missing space here

File: lib/patpass/patpass_comercio/inscription/InscriptionStartResponse.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    public $token;
+    public $url;
+
+    /**
+     * InscriptionStartResponse constructor.
+     * @param $token
+     */
+    public function __construct($json)
+    {
+        $token = $json["token"];
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/inscription/InscriptionStartResponse.php]: PHP throws if 'token' is not set on this $json

File: lib/webpay_rest/webpay_plus/exceptions/TransactionStatusException.php [hunk 6]
```diff
@@ -7,7 +7,7 @@
 use Transbank\Webpay\Exceptions\WebpayException;
 
 /**
- * Class TransactionStatusException
+ * Class InscriptionStatusException
```
Comment (eugenioLopezRamos) [inline lib/webpay_rest/webpay_plus/exceptions/TransactionStatusException.php]: I believe this change is wrong

File: tests/TransactionTest.php [hunk 7]
```diff
@@ -183,7 +183,7 @@ public function testTransactionCreationWorksWithOptions()
 
     public function testTransactionCommitWorks()
     {
-        // Setting commerce data
+        // Setting patpass_comercio data
```
Comment (eugenioLopezRamos) [inline tests/TransactionTest.php]: This change is wrong, these tests are for Onepay

File: tests/TransactionTest.php [hunk 8]
```diff
@@ -225,7 +225,7 @@ public function testTransactionCommitRaisesWhenResponseIsNull()
         $reflectedHttpClient->setAccessible(true);
         $reflectedHttpClient->setValue($httpClientStub);
 
-        // Setting commerce data
+        // Setting patpass_comercio data
```
Comment (eugenioLopezRamos) [inline tests/TransactionTest.php]: This change is wrong, this test is for Onepay

File: tests/TransactionTest.php [hunk 9]
```diff
@@ -316,7 +316,7 @@ public function testTransactionCommitRaisesWhenResponseSignatureIsNotValid()
...
```
Comment (eugenioLopezRamos) [inline tests/TransactionTest.php]: This change is wrong, this test is for Onepay

File: lib/patpass/patpass_comercio/Inscription.php [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+        if ($httpCode != 200 && $httpCode != 204) {
+            $reason = $httpResponse->getReasonPhrase();
+            $message = "Could not obtain a response from the service: $reason (HTTP code $httpCode)";
+            $body = json_decode($httpResponse->getBody(), true);
+            if (isset($body["error_message"])) {
+                $tbkErrorMessage = $body["error_message"];
+                $message = "$message. Details: $tbkErrorMessage";
+            }
+
+            throw new InscriptionStartException($message, -1);
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/Inscription.php]: The `-1`should be `$httpCode`

File: lib/patpass/patpass_comercio/Inscription.php [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        if ($httpCode != 200 && $httpCode != 204) {
+            $reason = $httpResponse->getReasonPhrase();
+            $message = "Could not obtain a response from the service: $reason (HTTP code $httpCode)";
+            $body = json_decode($httpResponse->getBody(), true);
+
+            if (isset($body["error_message"])) {
+                $tbkErrorMessage = $body["error_message"];
+                $message = "$message. Details: $tbkErrorMessage";
+            }
+            throw new InscriptionFinishException($message, -1);
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/Inscription.php]: The `-1` should be `$httpCode`

File: lib/patpass/patpass_comercio/Inscription.php [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        if ($httpCode != 200 && $httpCode != 204) {
+            $reason = $httpResponse->getReasonPhrase();
+            $message = "Could not obtain a response from the service: $reason (HTTP code $httpCode)";
+            $body = json_decode($httpResponse->getBody(), true);
+
+            if (isset($body["error_message"])) {
+                $tbkErrorMessage = $body["error_message"];
+                $message = "$message. Details: $tbkErrorMessage";
+            }
+            throw new InscriptionStatusException($message, -1);
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/Inscription.php]: The `-1` should be `$httpCode`

File: lib/patpass/patpass_comercio/exceptions/InscriptionStatusException.php [hunk 13]
```diff
@@ -0,0 +1,14 @@
+<?php
+namespace Transbank\Patpass\PatpassComercio\Exceptions;
+
+use Transbank\Patpass\Exceptions\PatpassException;
+
+class InscriptionStatusException extends PatpassException {
+    const DEFAULT_MESSAGE = 'Could not get Transaction status. Please verify given parameters';
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/exceptions/InscriptionStatusException.php]: The default message needs to be changed :(

File: lib/webpay_rest/Options.php [hunk 14]
```diff
@@ -37,7 +37,7 @@ class Options
      */
     public $apiKey = null;
     /**
-     * @var string $commerceCode Your commerce code, given by Transbank. Sent as
+     * @var string $commerceCode Your patpass_comercio code, given by Transbank. Sent as
```
Comment (eugenioLopezRamos) [inline lib/webpay_rest/Options.php]: This change is wrong

File: composer.json [hunk 15]
```diff
@@ -20,12 +20,16 @@
...
                   "Transbank\\Webpay\\Exceptions\\":["lib/webpay_rest/exceptions"],
-                  "Transbank\\Webpay\\WebpayPlus\\":["lib/webpay_rest/webpay_plus"], "Transbank\\Webpay\\WebpayPlus\\Exceptions\\":["lib/webpay_rest/webpay_plus/exceptions"], "Transbank\\Webpay\\Oneclick\\":["lib/webpay_rest/oneclick", "lib/webpay_rest/oneclick/inscription", "lib/webpay_rest/oneclick/mall"],
-        "Transbank\\Webpay\\Oneclick\\Exceptions\\":["lib/webpay_rest/oneclick/exceptions", "lib/webpay_rest/oneclick/mall/exceptions"],
+                  "Transbank\\Webpay\\WebpayPlus\\":["lib/webpay_rest/webpay_plus"],
+                  "Transbank\\Webpay\\WebpayPlus\\Exceptions\\":["lib/webpay_rest/webpay_plus/exceptions"],
+                  "Transbank\\Webpay\\Oneclick\\":["lib/webpay_rest/oneclick", "lib/webpay_rest/oneclick/inscription", "lib/webpay_rest/oneclick/mall"],
+                  "Transbank\\Webpay\\Oneclick\\Exceptions\\":["lib/webpay_rest/oneclick/exceptions", "lib/webpay_rest/oneclick/mall/exceptions"],
                   "Transbank\\Patpass\\":[ "lib/patpass","lib/patpass/patpass_by_webpay"],
```
Comment (eugenioLopezRamos) [inline composer.json]: You need to add the `patpass_comercio`directory here so your `Transbank\\Patpass\\PatpassComercio` class works.

File: lib/webpay_rest/webpay_plus/exceptions/TransactionStatusException.php (hunk 6 above)
Comment (a1b4) [reply lib/webpay_rest/webpay_plus/exceptions/TransactionStatusException.php]: @TheYakuza Eugenio is right, you should not change que Webpay Class name unless you have a very good reason

File: composer.json [hunk 16]
```diff
@@ -20,12 +20,16 @@
...
                   "Transbank\\Patpass\\PatpassByWebpay\\":["lib/patpass/patpass_by_webpay"],
-                  "Transbank\\Patpass\\Exceptions\\":["lib/patpass/exceptions"],
                   "Transbank\\Patpass\\PatpassByWebpay\\Exceptions\\":["lib/patpass/patpass_by_webpay/exceptions"],
+        "Transbank\\Patpass\\Exceptions\\": ["lib/patpass/exceptions"],
+        "Transbank\\Patpass\\PatpassComercio\\": ["lib/patpass/patpass_comercio", "lib/patpass/patpass_comercio/inscription"],
+        "Transbank\\Patpass\\PatpassComercio\\Exceptions\\": ["lib/patpass/patpass_commercio/exceptions"],
```
Comment (eugenioLopezRamos) [inline composer.json]: There is a typo here `patpass_commercio` should be `patpass_comercio`

File: lib/patpass/Options.php [hunk 17]
```diff
@@ -17,6 +17,7 @@ class Options
     const DEFAULT_API_KEY = '579B532A7440BB0C9079DED94D31EA1615BACEB56610332264630D42D0A36B1C';
     const DEFAULT_INTEGRATION_TYPE = "TEST";
     const DEFAULT_PATPASS_BY_WEBPAY_COMMERCE_CODE = '597055555550';
+    const DEFAULT_PATPASS_COMMERCE_COMMERCE_CODE = '';
```
Comment (eugenioLopezRamos) [inline lib/patpass/Options.php]: Is this correct?

File: lib/webpay_rest/Options.php (hunk 14 above)
Comment (a1b4) [reply lib/webpay_rest/Options.php]: @TheYakuza Eugenio is right, you should not change Webpay definitions unless you have a very good reason

File: lib/patpass/patpass_comercio/PatpassComercio.php [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        "LIVE" => "https://webpay3g.transbank.cl/",
+        "TEST" => "https://webpay3gint.transbank.cl/",
+        "MOCK" => ""
+        ];
+
+    public static $httpClient = null;
+    private static $apiKey = Options::DEFAULT_PATPASS_API_KEY;
+    private static $commerceCode = Options::DEFAULT_PATPASS_COMMERCE_COMMERCE_CODE;
+    private static $integrationType = Options::DEFAULT_PATPASS_INTEGRATION_TYPE;
+    private static $name = Options::DEFAULT_PATPASS_NAME;
```
Comment (eugenioLopezRamos) [inline lib/patpass/patpass_comercio/PatpassComercio.php]: I assume this is the `commerce` name? If so, for consistency you should call it `commerce_name` (like  the email)

File: lib/patpass/patpass_comercio/Inscription.php (hunk 1 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/Inscription.php]: Change list of vars to an array

File: lib/patpass/patpass_comercio/exceptions/InscriptionFinishException.php (hunk 2 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/exceptions/InscriptionFinishException.php]: fixed at eb51c37

File: lib/patpass/patpass_comercio/inscription/InscriptionFinishResponse.php (hunk 3 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/inscription/InscriptionFinishResponse.php]: There're not docs with this information

File: lib/patpass/patpass_comercio/inscription/InscriptionStatusResponse.php (hunk 4 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/inscription/InscriptionStatusResponse.php]: fixed at 5f2b809

File: lib/webpay_rest/webpay_plus/exceptions/TransactionStatusException.php (hunk 6 above)
Comment (TheYakuza) [reply lib/webpay_rest/webpay_plus/exceptions/TransactionStatusException.php]: fixed at 6364aad

File: tests/TransactionTest.php (hunk 8 above)
Comment (TheYakuza) [reply tests/TransactionTest.php]: fixed at 6364aad

File: tests/TransactionTest.php (hunk 9 above)
Comment (TheYakuza) [reply tests/TransactionTest.php]: fixed at 6364aad

File: lib/patpass/patpass_comercio/Inscription.php (hunk 10 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/Inscription.php]: fixed at 8320351

File: lib/patpass/patpass_comercio/Inscription.php (hunk 11 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/Inscription.php]: fixed at 8320351

File: lib/patpass/patpass_comercio/Inscription.php (hunk 12 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/Inscription.php]: fixed at 8320351

File: lib/patpass/patpass_comercio/Inscription.php (hunk 11 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/Inscription.php]: fixed at 8320351

File: lib/patpass/patpass_comercio/inscription/InscriptionStartResponse.php (hunk 5 above)
Comment (TheYakuza) [reply lib/patpass/patpass_comercio/inscription/InscriptionStartResponse.php]: fixed at f3fc8b3

File: lib/patpass/patpass_comercio/inscription/InscriptionFinishResponse.php (hunk 3 above)
Comment (eugenioLopezRamos) [reply lib/patpass/patpass_comercio/inscription/InscriptionFinishResponse.php]: You could use whatever Transbank returns from the service

File: lib/patpass/Options.php [hunk 19]
```diff
@@ -3,116 +3,97 @@
...
 class Options
 {
-    /**
-     * Default API key (which is sent as a header when making requests to Transbank
-     * on a field called "Tbk-Api-Key-Secret")
-     */
-    const DEFAULT_API_KEY = '579B532A7440BB0C9079DED94D31EA1615BACEB56610332264630D42D0A36B1C';
-    const DEFAULT_INTEGRATION_TYPE = "TEST";
-    const DEFAULT_PATPASS_BY_WEBPAY_COMMERCE_CODE = '597055555550';
+    const DEFAULT_PATPASS_API_KEY = "giPmNIQIabX8KADNqJht   ";
```
Comment (eugenioLopezRamos) [inline lib/patpass/Options.php]: Are the spaces at the end correct?

File: lib/patpass/Options.php [hunk 20]
```diff
@@ -3,116 +3,97 @@
...
 
-    /**
-     * @return Options Return an instance of Options with default values
-     * configured
-     */
     public static function defaultConfig()
     {
-        return new Options(self::DEFAULT_API_KEY,
-            self::DEFAULT_PATPASS_BY_WEBPAY_COMMERCE_CODE);
+        return new Options(
```
Comment (eugenioLopezRamos) [inline lib/patpass/Options.php]: Yeah, this is not gonna work. We need to make this work for both Patpass by webpay and patpass comercio. Probably should be 2 different methods...

File: lib/patpass/Options.php [hunk 21]
```diff
@@ -3,116 +3,97 @@
...
  */
 class Options
 {
-    /**
-     * Default API key (which is sent as a header when making requests to Transbank
-     * on a field called "Tbk-Api-Key-Secret")
-     */
-    const DEFAULT_API_KEY = '579B532A7440BB0C9079DED94D31EA1615BACEB56610332264630D42D0A36B1C';
-    const DEFAULT_INTEGRATION_TYPE = "TEST";
-    const DEFAULT_PATPASS_BY_WEBPAY_COMMERCE_CODE = '597055555550';
```
Comment (eugenioLopezRamos) [inline lib/patpass/Options.php]: Why is this being deleted?

**Comment by TheYakuza** [comment]
close PR to prevent conflicts whit branch feat/add-rest-support

**Pull request closed without merging**